### PR TITLE
SNOW-1665417 schema evolution primitives

### PR DESCRIPTION
**Pull request by sfc-gh-wtrefon**
# Overview

SNOW-1665417

## Pre-review checklist
- [ ] This change should be part of a Behavior Change Release. See [go/behavior-change](http://go/behavior-change).
- [ ] This change has passed Merge gate tests
- [ ] Snowpipe Changes
- [ ] Snowpipe Streaming Changes
- [ ] This change is TEST-ONLY
- [ ] This change is README/Javadocs only
- [ ] This change is protected by a config parameter <PARAMETER_NAME> eg `snowflake.ingestion.method`.
    - [ ] `Yes` - Added end to end and Unit Tests. 
    - [ ] `No` - Suggest why it is not param protected
- [ ] Is his change protected by parameter <PARAMETER_NAME> on the server side?
    - [ ] The parameter/feature is not yet active in production (partial rollout or PrPr, see [Changes for Unreleased Features and Fixes](http://go/ppp-prpr)).
    - [ ] If there is an issue, it can be safely mitigated by turning the parameter off. This is also verified by a test (See [go/ppp](http://go/ppp)).

### CODE REVIEW

File: src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java [hunk 1]
```diff
@@ -1,62 +1,235 @@
...
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Disabled;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
 
-public class IcebergIngestionIT {
-  private static final String tableName = "tbl1";
+public class IcebergIngestionIT extends BaseIcebergIT {
```
Comment (sfc-gh-wtrefon) [inline src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java:36]: more tests cases and probably some refactoring will come in next PRs, this PR is already quite big

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/snowflake/SnowflakeColumnTypeMapper.java [hunk 2]
```diff
@@ -57,4 +69,33 @@ public String mapToColumnType(Schema.Type kafkaType, String schemaName) {
...
+    } else if (value.isBoolean()) {
+      return BOOLEAN;
+    } else if (value.isBinary()) {
+      return BYTES;
+    } else if (value.isArray()) {
+      return ARRAY;
+    } else if (value.isObject()) {
+      return STRUCT;
+    } else {
+      return null;
```
Comment (sfc-gh-akowalczyk) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/snowflake/SnowflakeColumnTypeMapper.java:98]: We do not expect to hit it ever, do we? Maybe we should log at least a warning here unless we want to throw an exception and break the execution

Comment (sfc-gh-wtrefon) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/snowflake/SnowflakeColumnTypeMapper.java:98]: this is old code, just moved it. If we hit null here then we will get `throw new IllegalArgumentException("Arrays, struct and map not supported!");`
in `mapToColumnType` method

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/snowflake/SnowflakeColumnTypeMapper.java [hunk 3]
```diff
@@ -57,4 +69,33 @@ public String mapToColumnType(Schema.Type kafkaType, String schemaName) {
         return "VARIANT";
     }
   }
+
+  public Schema.Type mapJsonNodeTypeToKafkaType(JsonNode value) {
```
Comment (sfc-gh-mbobowski) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/snowflake/SnowflakeColumnTypeMapper.java:73]: `@Override`

File: src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java [hunk 4]
```diff
@@ -1,62 +1,231 @@
...
+
+  private final String simpleRecordJson = "{\"simple\": \"extra field\"}";
+
+  private static final String primitiveJson =
+      "{\n"
+          + "  \"id_int8\": 0,\n"
+          + "  \"id_int16\": 42,\n"
+          + "  \"id_int32\": 42,\n"
+          + "  \"id_int64\": 42,\n"
+          + "  \"description\": \"dogs are the best\",\n"
```
Comment (sfc-gh-mbobowski) [inline src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java:53]: @sfc-gh-lshcharbaty may request change on that

File: src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java [hunk 5]
```diff
@@ -1,62 +1,231 @@
...
-            .setProperties(props)
-            // .setIsIceberg(true)
-            .build();
+  private SnowflakeSinkService service;
+
+  private final String simpleRecordJson = "{\"simple\": \"extra field\"}";
+
+  private static final String primitiveJson =
+      "{\n"
+          + "  \"id_int8\": 0,\n"
```
Comment (sfc-gh-mbobowski) [inline src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java:49]: Maybe we can skip `\n`?

Comment (sfc-gh-wtrefon) [reply src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java:49]: good idea, it does not matter to jackson anyways

File: src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java [hunk 6]
```diff
@@ -1,62 +1,231 @@
...
-            .setOnErrorOption(OpenChannelRequest.OnErrorOption.CONTINUE)
+  @BeforeEach
+  public void setUp() {
+    tableName = TestUtils.randomTableName();
+    topic = tableName;
+    topicPartition = new TopicPartition(topic, PARTITION);
+    Map<String, String> config = getConfForStreaming();
+    SnowflakeSinkConnectorConfig.setDefaultValues(config);
+    config.put(ICEBERG_ENABLED, "TRUE");
+    config.put(ENABLE_SCHEMATIZATION_CONFIG, "TRUE");
```
Comment (sfc-gh-mbobowski) [inline src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java:112]: nit: Separating schematization and non-schematization ITs into separate files may be a good idea. Doesn't have to be done within this PR.

File: src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java [hunk 7]
```diff
@@ -1,62 +1,231 @@
...
+
+    assertThat(rows).containsExactlyInAnyOrder(expectedSchema);
+
+    // resend and store same record without any issues now
+    service.insert(Collections.singletonList(record));
+    TestUtils.assertWithRetry(() -> service.getOffset(topicPartition) == 1, 20, 5);
+
+    // and another record with same schema
+    service.insert(Collections.singletonList(createKafkaRecord(message, 1, withSchema)));
+    TestUtils.assertWithRetry(() -> service.getOffset(topicPartition) == 2, 20, 5);
```
Comment (sfc-gh-mbobowski) [inline src/test/java/com/snowflake/kafka/connector/streaming/iceberg/IcebergIngestionIT.java:170]: nit: having overloaded `assertWithRetry` without interval details might be a good option

**Review by sfc-gh-mbobowski — APPROVED** [review]
Big step forward!

**Pull request merged**